File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    
    visited = [[False] * m for _ in range(n)]
    dist = [[-1] * m for _ in range(n)]
    
    q = deque()
    visited[0][0] = True
    dist[0][0] = 1
    q.append((0, 0))
    
    dir = [[-1, 0], [0, 1], [1, 0], [0, -1]]
    
    
    while q:
        vy, vx = q.popleft()
        
        if vy == n and vx == m:
            break
        
        for i in range(4):
            ny = vy + dir[i][0]
            nx = vx + dir[i][1]
            
            if 0<=ny<n and 0<=nx<m and not visited[ny][nx] and maps[ny][nx] == 1:
                visited[ny][nx] = True
                dist[ny][nx] = dist[vy][vx]+1
                q.append((ny, nx))
                
    
    
    answer = dist[n-1][m-1]
    
    
    return answer

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    class Node{
        int y, x;
        boolean visited;
        int dist;
        
        void setNode(int y, int x){
            this.y = y;
            this.x = x;
            dist = -1;
            visited= false;
        }
    }
    
    static int dir[][] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
    
    public boolean canVisit(int ny, int nx, int n, int m){
        return ny>=0 && ny<n && nx>=0 && nx<m;
    }
    
    public int solution(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        
        Node node[][] = new Node[101][101];
        
        for(int i = 0; i<n; i++){
            for(int j = 0; j<m; j++){
                node[i][j] = new Node();
                node[i][j].setNode(i, j);
            }
        }
        
        
        Queue<Node> q = new LinkedList<>();
        
        Node v;
        
        node[0][0].visited = true;
        node[0][0].dist = 1;
        
        q.add(node[0][0]);
        
        while(!q.isEmpty()){
            v = q.element();
            q.remove();
            
            int ny, nx;
            
            if(v.y == n-1 && v.x == m-1){
                break;
            }
            
            for(int i = 0; i<4; i++){
                ny = v.y + dir[i][0];
                nx = v.x + dir[i][1];
                
                if(canVisit(ny, nx, n, m) && maps[ny][nx] == 1 && !node[ny][nx].visited){
                    node[ny][nx].visited = true;
                    node[ny][nx].dist = v.dist + 1;
                    q.add(node[ny][nx]);
                }
            }
        }
        
        int answer = node[n-1][m-1].dist;
        return answer;
    }
}